src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,89 @@
+                throw new ArgumentException("Handle of a NativeLibrary can't be set to IntPtr.Zero.");
+            Handle = handle;
+        }
+
+        static NativeLibrary()
+        {
+            _assemblyToCallbackMap = new ConditionalWeakTable<Assembly, Func<LoadNativeLibraryArgs, NativeLibrary>>();
+        }
+
+        /// <exception cref="System.Runtime.InteropServices.CallbackAlreadyRegistered">Thrown when there is already a callback registered for the specified assembly.</exception>
+        public static void RegisterNativeLibraryLoadCallback(Assembly assembly, Func<LoadNativeLibraryArgs, NativeLibrary> callback)
+        {
+
+            if (_assemblyToCallbackMap.TryGetValue(assembly, out Func<LoadNativeLibraryArgs, NativeLibrary> previousCallback))
+            {
+                throw new CallbackAlreadyRegisteredException("Callback for " + assembly.GetName().Name + " has already been registered.");
+            }
+            else
+            {
+                _assemblyToCallbackMap.Add(assembly, callback);
+            }
+        }
+
+        // A wrapper function for Load(string libraryName, DllImportSearchPath dllImportSearchPath, Assembly assembly)
+        public static NativeLibrary Load(string libraryName)
+        {
+            return Load(libraryName, DllImportSearchPath.LegacyBehavior, null);
+        }
+
+        public static NativeLibrary Load(string libraryName, DllImportSearchPath dllImportSearchPath, Assembly assembly)
+        {
+            RuntimeAssembly assemblyAsRuntimeAssembly = null;
+            if (assembly != null)
+            {
+                assemblyAsRuntimeAssembly = (RuntimeAssembly)assembly;
+            }
+
+            IntPtr hmodule = LoadLibrary(assemblyAsRuntimeAssembly, libraryName, (int)dllImportSearchPath);
+
+            NativeLibrary loadedLibrary = new NativeLibrary(libraryName, hmodule);
+
+            return loadedLibrary;
+        }
+
+        // Calls NativeLibrary::LoadLibrary unmanaged methode
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        internal static extern IntPtr LoadLibrary(RuntimeAssembly assembly, string libraryName, int dllImportSearchPathFlag);
+
+        // Todo: optimize for methods that do not have callbacks
+        internal static IntPtr LoadLibraryCallback(string libraryName, uint dllImportSearchPathUint, Assembly assembly)
+        {
+            DllImportSearchPath dllImportSearchPath = (DllImportSearchPath)dllImportSearchPathUint;
+
+            bool callbackFound = _assemblyToCallbackMap.TryGetValue(assembly, out Func<LoadNativeLibraryArgs, NativeLibrary> callback);
+
+            if (callbackFound)
+            {
+                LoadNativeLibraryArgs loadNativeLibraryArgs = new LoadNativeLibraryArgs(libraryName, dllImportSearchPath, assembly);
+                NativeLibrary nativeLibrary = callback(loadNativeLibraryArgs);
+                if(nativeLibrary != null) return nativeLibrary.Handle;
```

Comment:
nit: do you need if(nativeLibrary != null)?

Reply:
What if a callback returns null instead of an instance of the NativeLibrary? It may happen for callbacks implemented by users and I think we don't want to throw NullReferenceException then.

Reply:
You are right. We do need this check.